### DISCUSSION TITLE
Signing SAML requests with a different certificate for each provider

### DISCUSSION THREAD

**Discussion by Ketec** [keycloak/keycloak#31580] [Q&A]
A similar question exists from a year ago, but the answer does not seem to apply to the current (24/25 quarkus) keycloak versions, i suspect it's for some custom build.
https://github.com/keycloak/keycloak/discussions/20153
No extension examples either that i can find for new KC versions.

Various government systems (and many larger companies) specifically require their SSO request to be signed with a specific certificate from their CA.
But currently, keycloak only has one global realm level signing key.

Multiple realms are not an option - the users still log into one app - and all interactions, token introspection etc are for one realm.
Creating many realms is not maintainable and won't perform well (+ various token context issues).

Extending the SAML IDP with SPI to create a custom is impossible. The new React based admin UI uses hard coded provider id-s and the extended SPI has no fields.
And there is no way to add additional fields (which would have been doable with Freemaker).

Has anyone found a way to correctly extend keycloak SAML provider?

**Comment by harthorst** [comment]
I found this approach which works well: https://gist.github.com/Szer/fda52c4afaaca9c3a4aee35b7870d2ed

But I'd like to discuss whether this is the right way. Opinions are welcome.
